### PR TITLE
Failed tests counter fix

### PR DESCRIPTION
**Pull request by AndreiZaitcev**
Updates a counter when a test run fails, addresses potential multithreading issues plus cleans up.

### CODE REVIEW

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java [hunk 1]
```diff
@@ -1,69 +1,78 @@
...
-
-    private SetMultimap<Pool, TestCaseEventCounter> map = HashMultimap.<Pool, TestCaseEventCounter>create();
+    private final SetMultimap<Pool, TestCaseEventCounter> testCaseCounters = synchronizedSetMultimap(create());
 
     @Override
     public void record(Pool pool, TestCaseEvent testCaseEvent) {
-        if (!map.containsKey(pool)) {
-            map.put(pool, createNew(testCaseEvent));
+        if (!testCaseCounters.containsKey(pool)) {
+            testCaseCounters.put(pool, createNew(testCaseEvent));
```
Comment (edwardharks) [inline fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: Shouldn't this go inside a synchronized block?

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java [hunk 2]
```diff
@@ -1,69 +1,78 @@
...
         }
     }
 
     @Override
     public int getCount(Pool pool, TestCaseEvent testCaseEvent) {
-        if (map.containsKey(pool)) {
-            return from(map.get(pool))
-                    .firstMatch(isSameTestCase(testCaseEvent)).or(TestCaseEventCounter.EMPTY)
-                    .getCount();
+        if (testCaseCounters.containsKey(pool)) {
```
Comment (edwardharks) [inline fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: Also this should be synchronized too

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java [hunk 3]
```diff
@@ -1,69 +1,78 @@
...
+import static com.google.common.collect.HashMultimap.create;
+import static com.google.common.collect.Multimaps.synchronizedSetMultimap;
 
 /**
  * Class that keeps track of the number of times each testCase is executed for device.
  */
 public class PoolTestCaseFailureAccumulator implements PoolTestCaseAccumulator {
-
-    private SetMultimap<Pool, TestCaseEventCounter> map = HashMultimap.<Pool, TestCaseEventCounter>create();
+    private final SetMultimap<Pool, TestCaseEventCounter> testCaseCounters = synchronizedSetMultimap(create());
```
Comment (edwardharks) [inline fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: Do you need a synchronizedSetMultimap if everything is inside a synchronized block?

File: fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java [hunk 4]
```diff
@@ -26,36 +26,32 @@
...
     private final Device device;
-    private final TestCaseEvent currentTestCaseEvent;
-    private final TestRetryer testRetryer;
+    private final FailedTestScheduler failedTestScheduler;
     private final Pool pool;
     private final DeviceTestFilesCleaner deviceTestFilesCleaner;
-    private TestIdentifier startedTest;
-    private TestIdentifier failedTest;
+    private TestCaseEvent startedTest;
+    private TestCaseEvent failedTest;
```
Comment (edwardharks) [inline fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java]: Why is this changed from TestIdentifier to TestCaseEvent?

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java (hunk 1 above)
Comment (AndreiZaitcev) [reply fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: No if I understand correctly. We have to wrap it inside a synchronized block only when we iterate over this collection. More info here https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multimaps.html#synchronizedMultimap(com.google.common.collect.Multimap)

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java (hunk 2 above)
Comment (AndreiZaitcev) [reply fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: Please check the answer above.

File: fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java (hunk 3 above)
Comment (AndreiZaitcev) [reply fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java]: Yes, please check docs for Multimaps.synchronizedMultimap https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multimaps.html#synchronizedMultimap(com.google.common.collect.Multimap)

File: fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java (hunk 4 above)
Comment (AndreiZaitcev) [reply fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java]: I refactored FailedTestScheduler to use our custom model `TestCaseEvent` instead of `TestIdentifier` which is coming from ddmlib.

Comment (edwardharks) [reply fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java]: 🥇
